Resource deletion page

siovan-ssc opened the pull request:
# Pull Request

## Description
<!-- A brief description of what this PR does -->
Adding resource deletion functionality. Mostly just want eyes on the code, not sure how to do the resource deletion queuing.
![image](https://github.com/user-attachments/assets/b57e1606-fd0a-4231-8d63-4f8a05cbb2dc)
![image](https://github.com/user-attachments/assets/2caaaf27-e715-4a27-a451-4ea1a2a23ff3)


## Related Issues
<!-- List any related issues or tickets -->

## Changes
<!-- List the key changes in this PR -->

- Adding "delete tool" button to toolbox resources + functionality

## Testing
<!-- Describe the tests that were run or any QA steps that were taken -->

- Visual

## Checklist
<!-- Ensure the following tasks are complete -->

- [X] Code follows dotnet coding standards
- [X] Documentation updated (if necessary)
- [X] Tests added/updated (if necessary)

siovan-ssc committed:
Basic page navigation

siovan-ssc committed:
Delete implementation

Still issues, pushing in case we want scheduling at some point.

siovan-ssc committed:
Updates to delete functionality

siovan-ssc committed:
Cleaning up

siovan-ssc committed:
Removing test string

davidreneuw commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -105,45 +123,46 @@
+        { TerraformTemplate.AzureAppService, ProjectResourceStatus.Available },
+        { TerraformTemplate.AzurePostgres, ProjectResourceStatus.Available },
+        { TerraformTemplate.AzureArcGis, ProjectResourceStatus.UnderDevelopment },
+        { TerraformTemplate.AzureAPI, ProjectResourceStatus.UnderDevelopment }
```
You should make use of AvailabilityStatus instead of ProjectResourceStatus here

siovan-ssc replied:
I was going to make these changes but then I realized it would be a dictionary of <string, AvailabilityStatus> which would cause issues with a lot of the rest of the code since it's comparing/using ProjectResourceStatus. i think the whole thing would need to be cleaned up/refactored to support this but I can def give it a shot if it's of particular importance!

yjmrobert commented on Portal/src/Datahub.Infrastructure/Services/ResourceMessagingService.cs:
```diff
@@ -39,6 +39,28 @@ public async Task SendToTerraformDeleteQueue(WorkspaceDefinition project, int pr
+        if (resourceToDeleted != null && projectChanged != null)
+        {
+            resourceToDeleted.UpdatedAt = DateTime.UtcNow;
+            projectChanged.Resources.Remove(resourceToDeleted);
```
We don't want to actually delete the database record of it here but rather do a "soft delete" instead.

Either way it's outside the scope of this PR but just a heads up! 😄

yjmrobert reviewed (dismissed):
Great work, I'm a big fan of the enter acronym to delete!

yjmrobert committed:
Merge branch 'develop' into sg/resource-deletion-page

siovan-ssc committed:
Commenting out enum

siovan-ssc committed:
Merge branch 'sg/resource-deletion-page' of https://github.com/ssc-sp/datahub-portal into sg/resource-deletion-page

yjmrobert committed:
Merge branch 'develop' into sg/resource-deletion-page

siovan-ssc committed:
Commenting out deleteresourcequeue

davidreneuw commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -168,59 +187,200 @@
+        var _resourceCheck = await ctx.Project_Resources2
+            .FirstOrDefaultAsync(r => r.ResourceType.EndsWith(tool));
+        _resourceCheck.Status = ProjectResourceStatus.PendingDeletion.ToString();
+
```
You should use `TerraformTemplate.GetTerraformServiceType(string templateName)` to get the value that is used in the `ResourceType` column of the `Project_Resources`, instead of comparing the ends of the strings.
E.g:
```csharp
var resourceType = TerraformTemplate.GetTerraformServiceType(tool);
using var ctx = await _contextFactory.CreateDbContextAsync();
var _resourceCheck = await ctx.Project_Resources2.FirstOrDefaultAsync(r => r.ResourceType == resourceType);
```

davidreneuw commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -168,59 +187,200 @@
+        await SendEmailAsync(async () =>
+        {
+            await _emailService.SendToRecipients("ssc.datahub.donotreply-nepasrepondre.datahub.spc@ssc-spc.gc.ca", recipients, subject, body);
+        });
```
At the top of the page, you should inject:

```csharp
@inject DatahubPortalConfiguration _datahubPortalConfiguration
```

Then here, in the email service, you should use the local configurations instead of manually writing the email:

```csharp
await _emailService.SendToRecipients(_datahubPortalConfiguration.EmailNotification.SenderAddress, recipients, subject, body);
```

This way if we ever change the email we won't have to go around the whole codebase

davidreneuw commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -168,59 +187,200 @@
+        return ProjectResourceStatus.Error;
+    }
+
+    private async void ToolDeletion()
```
You should be careful, `ToolDeletion()` is used in `OnInitializedAsync()`, and puts a deletion message on the queue if any workspace tool's status is "SendingDeletion". A user could unknowingly spam the resource deletion queue with invalid messages by refreshing the page

davidreneuw reviewed (dismissed):
Looks good!

siovan-ssc committed:
Initial changes based on David's suggestions

yjmrobert committed:
Merge branch 'develop' into sg/resource-deletion-page

yjmrobert committed:
🔧 Add project resource status with detailed documentation

Detailed XML documentation was added to describe the possible values and their significance. Removed the obsolete ProjectResourceStatus enum to clean up the code.

yjmrobert committed:
🔧 Add new Terraform status: DeleteInProgress

Introduce a new status constant 'DeleteInProgress' to track deletion processes. This will help in better monitoring and managing the current state of Terraform operations.

yjmrobert committed:
🔧 Add ToolboxButton component and refactor toolbox logic

Introduce a new ToolboxButton component to manage the status and actions of toolbox resources. Additionally, refactor the workspace toolbox and deletion pages for improved status handling and component reusability, ensuring clearer and more maintainable code.

yjmrobert committed:
📜 Add TODO marker for future delete logic

Inserted a placeholder comment indicating where delete logic should be added in `RequestManagementService.cs`. This serves as a reminder to implement the necessary functionality in upcoming updates.

yjmrobert committed:
🔧 Add initialization loading indicator to ToolboxPage

Introduced a loading indicator during the initialization phase of the `WorkspaceToolboxPage`. This enhancement improves user experience by displaying a loading message until all necessary data is fully loaded, ensuring a smoother interface interaction.

Sean-Stilwell commented on Portal/src/Datahub.Portal/Pages/Workspace/Toolbox/WorkspaceToolboxPage.razor:
```diff
@@ -152,76 +148,94 @@
+            var exists = await ctx.Project_Resources2.AnyAsync(r => r.ResourceType == resourceType);
+            if (exists)
+            {
+                _snackbar.Add(Localizer["{0} is already added to your workspace", GetLabel(resource)], Severity.Info);
```
This is currently displaying every time I try to add a resource to a workspace

yjmrobert replied:
It's a bug I fixed it in the subsequent PR

yjmrobert replied:
Forgot to scope the check to the workspace, just goes against the whole DB

https://github.com/ssc-sp/datahub-portal/commit/afb718799a82283710091cbde4410e416ab2fdb6

ErikApption committed:
Merge branch 'develop' into sg/resource-deletion-page

davidreneuw committed:
Merge branch 'develop' into sg/resource-deletion-page

ErikApption committed:
Merge branch 'develop' into sg/resource-deletion-page

ErikApption committed:
Merge branch 'develop' into sg/resource-deletion-page

ErikApption committed:
Merge branch 'develop' into sg/resource-deletion-page
